src/v/coproc/pacemaker.cc

```diff
@@ -112,7 +114,24 @@ std::vector<errc> pacemaker::add_source(
       id, _shared_res, std::move(ctxs));
     const auto [itr, success] = _scripts.emplace(id, std::move(script_ctx));
     vassert(success, "Double coproc insert detected");
-    (void)itr->second->start();
+    vlog(coproclog.debug, "Adding source with id: {}", id);
+    (void)itr->second->start().handle_exception_type(
+      [this, id](const script_failed_exception& e) {
+          /// A script must be deregistered due to an internal script error.
+          /// The wasm engine determines the case, most likley the apply()
+          /// method has thrown or there is a syntax error within the script
+          /// itself.
+          vlog(coproclog.info, "Handling script_failed_exception: {}", e);
+          vassert(
+            id == e.get_id(),
+            "script_failed_handler id mismatch detected, observed: {} "
+            "expected: {}",
+            e.get_id(),
+            id);
+          return container().invoke_on_all([id](pacemaker& p) {
+              return p.remove_source(id).discard_result();
+          });
```

Comment:
and the handling here (as opposed to inside start()) is because the error bubbles up here where there is access to the pacemaker shards and can turn around and remove the source?

Reply:
yup thats correct